model/src/lib.rs

```diff
@@ -293,4 +370,69 @@ mod tests {
         let token = H160::from_low_u64_be(1);
         assert_eq!(TokenPair::new(token, token), None);
     }
+
+    #[test]
+    fn signature_typed_data() {
+        // copied from a gp-v2-contracts test
+        let domain_separator =
+            hex!("f8a1143d44c67470a791201b239ff6b0ecc8910aa9682bebd08145f5fd84722b");
+        let order = OrderCreation {
+            sell_token: hex!("0101010101010101010101010101010101010101").into(),
+            buy_token: hex!("0202020202020202020202020202020202020202").into(),
+            sell_amount: hex!("0303030303030303030303030303030303030303030303030303030303030303")
+                .into(),
+            buy_amount: hex!("0404040404040404040404040404040404040404040404040404040404040404")
+                .into(),
+            valid_to: 84215045,
+            app_data: 101058054,
+            fee_amount: hex!("0707070707070707070707070707070707070707070707070707070707070707")
+                .into(),
+            order_kind: OrderKind::Buy,
+            partially_fillable: true,
+            signature: Signature {
+                v: 0x1b,
+                r: hex!("159bad4cdce8e9eeb34f4450941e6e512cd2ceaba2809f6928ad91ce58700064").into(),
+                s: hex!("4cd8e52110ca7d1ba7493828bb811969115aff9d8358a5071bd2e7d15c1362bd").into(),
+            },
+        };
+        let expected_owner = hex!("70997970C51812dc3A010C7d01b50e0d17dc79C8");
+        let owner = order.validate_signature(&domain_separator).unwrap();
+        assert_eq!(owner, expected_owner.into());
+    }
+
+    #[test]
+    fn signature_message() {
+        // copied from a gp-v2-contracts test
+        let domain_separator =
+            hex!("f8a1143d44c67470a791201b239ff6b0ecc8910aa9682bebd08145f5fd84722b");
+        let order = OrderCreation {
+            sell_token: hex!("0101010101010101010101010101010101010101").into(),
+            buy_token: hex!("0202020202020202020202020202020202020202").into(),
+            sell_amount: hex!("0246ddf97976680000").as_ref().into(),
+            buy_amount: hex!("b98bc829a6f90000").as_ref().into(),
+            valid_to: 4294967295,
+            app_data: 0,
+            fee_amount: hex!("0de0b6b3a7640000").as_ref().into(),
+            order_kind: OrderKind::Sell,
+            partially_fillable: false,
+            signature: Signature {
+                v: 0x1b | 0x80,
+                r: hex!("97be7a77d916c99b39d2909c5fde76a82f4fd49868b18d000d93de2c59061602").into(),
+                s: hex!("68f04221fd6ac0e8aebfb048cf6c75ca882cd20c6e06aec094c278f8e44c1759").into(),
+            },
+        };
+        let expected_owner = hex!("70997970C51812dc3A010C7d01b50e0d17dc79C8");
+        let owner = order.validate_signature(&domain_separator).unwrap();
+        assert_eq!(owner, expected_owner.into());
+    }
+
+    #[test]
+    fn sign_self() {
+        let mut order = OrderCreation::default();
```

Comment:
I'd slightly prefer if we used UserOrder::default and in there create a real signature that is valid. Creating an invalid order creation and then mutating it into something allows to create semantically invalid data (not so important for the tests but also alluding to my general preference of checking signature at order creation time).

Reply:
Addressing this and 

> I think it would be nice if signature verification could happen on order creation so we disallow invalid orders to be created at all.

This type was intended to implement the serialization format. I can see some merit in having a checked `CorrectlySignedOrder` but I don't think enforcing that for this type meshes well with the serialization aspect. Enforcing that isn't as simple as it seems so I don't want to include it in this PR. I made an issue to discuss this in https://github.com/gnosis/gp-v2-services/issues/96 .